mkosi/__init__.py

```diff
@@ -236,28 +238,71 @@ def handler(signal: int, frame: FrameType) -> None:
             die("Interrupted")
 
 
+@contextlib.contextmanager
+def do_noop() -> Generator[None, None, None]:
+    yield
+
+
 # Borrowed from https://github.com/python/typeshed/blob/3d14016085aed8bcf0cf67e9e5a70790ce1ad8ea/stdlib/3/subprocess.pyi#L24
 _FILE = Union[None, int, IO[Any]]
 
 
+def spawn(
+    cmdline: List[str],
+    delay_interrupt: bool = True,
+    stdin: _FILE = None,
+    stdout: _FILE = None,
+    stderr: _FILE = None,
+    **kwargs: Any,
+) -> Popen:
+    if "run" in arg_debug:
+        MkosiPrinter.info("+ " + " ".join(shlex.quote(x) for x in cmdline))
+
+    if not stdout and not stderr:
+        # Unless explicit redirection is done, print all subprocess
+        # output on stderr, since we do so as well for mkosi's own
+        # output.
+        stdout = sys.stderr
+
+    if stdin is None:
+        stdin = subprocess.DEVNULL
+
+    cm = do_delay_interrupt if delay_interrupt else do_noop
+    try:
+        with cm():
+            return subprocess.Popen(cmdline, stdin=stdin, stdout=stdout, stderr=stderr, **kwargs)
+    except FileNotFoundError:
+        die(f"{cmdline[0]} not found in PATH.")
+
+
 def run(
     cmdline: List[str],
     check: bool = True,
+    delay_interrupt: bool = True,
+    input: Union[str, bytes, None] = None,
+    stdin: _FILE = None,
     stdout: _FILE = None,
     stderr: _FILE = None,
     **kwargs: Any,
 ) -> CompletedProcess:
     if "run" in arg_debug:
         MkosiPrinter.info("+ " + " ".join(shlex.quote(x) for x in cmdline))
 
+    if stdin is None and input is None:
+        stdin = subprocess.DEVNULL
+
     if not stdout and not stderr:
-        # Unless explicit redirection is done, print all subprocess output on stderr since we do so as well
-        # for mkosi's own output.
+        # Unless explicit redirection is done, print all subprocess
+        # output on stderr, since we do so as well for mkosi's own
+        # output.
         stdout = sys.stderr
 
+    cm = do_delay_interrupt if delay_interrupt else do_noop
     try:
-        with delay_interrupt():
-            return subprocess.run(cmdline, check=check, stdout=stdout, stderr=stderr, **kwargs)
+        with cm():
+            return subprocess.run(
+                cmdline, check=check, input=input, stdin=stdin, stdout=stdout, stderr=stderr, **kwargs
```

Comment:
This can potentially byte a user, since `input` and `stdin` cannot be non-`None` at the same time. It might make more sense to standardise on either in our wrapper.

Reply:
There's a check above, and I think that should be enough. Standardizing on one of them is not easy, because they have completely different types and semantics...

Reply:
True, the thought came from the above check being for both not being `None`, but in the end this will be caught in `subprocess.run` itself, so that's fine.